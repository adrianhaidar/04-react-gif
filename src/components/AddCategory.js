
import React, { useState } from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({ setCategories }) => {
    // Si se deja el useState vacío este sería undefined, por ello, se deja un string vacío
    const [inputValue, setinputValue] = useState(''); //useState('') es lo que va en el buscador

    const handleInputChange = (e) => {
        setinputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        // Debido a que ya no es muy usado la parte del refresh completo del navegador, que es un comportamiento por default de los formularios se utiliza la siguiente linea de código
        e.preventDefault();
        console.log('handleSubmit llamado');

        if (inputValue.trim().length > 1) {
            // Como no se tiene acceso a las categorias del componente padre hace uso un callback como se muestra, el inputValue seria el nuevo valor.
            setCategories((cats) => [inputValue, ...cats]);
            setinputValue('');
            //esto para que quede vacio el coso
        }
    };

    return (
        // Cuando se retorna na unica etiqueta agrupadora como es en este caso el form no es necesario agruparla dentro de un Fragment
        <form onSubmit={handleSubmit}>    
            <p>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    );
};
//el onSubmit es para que no haga refresh osea el handlesubmit con preventdefault
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
};

export default AddCategory;