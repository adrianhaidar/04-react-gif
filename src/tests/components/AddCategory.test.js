import '@testing-library/jest-dom';
import React from 'react';
import { shallow } from 'enzyme';
import AddCategory from '../../components/AddCategory';

describe('Pruebas en eln componente <AddCategory/>', () => {
    // Simula una función
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    // Esta funcón se ejecuta antes de cada prueba
    beforeEach(() => {
        jest.clearAllMocks(); // Limpia los atributos y simulaciones que se hicieron
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
    });

    test('Debe mostrar correctamente el componente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de cambiar el input', () => {
        const input = wrapper.find('input');
        const value = 'Hola mundo!!';

        // El segundo argumento que recibe el método simulate es el evento e, el cual tiene una propiedad llamada target que a su vez también tiene una propiedad llamada value
        input.simulate('change', { target: { value } });
        expect(wrapper.find('p').text().trim()).toBe(value);
    });

    test('No debe de postear la información on submit', () => {
        // Se envía como parametro la funcion preventDefault, esa forma es la manera corta de hacerlo como una función tradicional, aunque también se puede hacer con arrow functions
        wrapper.find('form').simulate('submit', { preventDefault() {} });
        expect(setCategories).not.toHaveBeenCalled();
    });

    test('Debe de llamr el setCategories y limpiar el inputChange', () => {
        const value = 'Hola mundo';
        const input = wrapper
            .find('input')
            .simulate('change', { target: { value } });
        wrapper.find('form').simulate('submit', { preventDefault() {} });
        expect(setCategories).toHaveBeenCalled();
        expect(input.prop('value')).toBe('');
    });
});