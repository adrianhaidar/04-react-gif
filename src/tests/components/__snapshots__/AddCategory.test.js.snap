// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en eln componente <AddCategory/> Debe mostrar correctamente el componente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;